### PR TITLE
Make ENV_LOCK reentrant

### PR DESCRIPTION
**Pull request by khwilliamson**
This is in preparation for future commits

### CODE REVIEW

File: perl.h [hunk 1]
```diff
@@ -7068,8 +7068,13 @@ typedef struct am_table_short AMTS;
 #endif
 
 #ifdef USE_THREADS
-#  define ENV_LOCK            PERL_WRITE_LOCK(&PL_env_mutex)
-#  define ENV_UNLOCK          PERL_WRITE_UNLOCK(&PL_env_mutex)
+#  define ENV_LOCK            PERL_REENTRANT_LOCK("env",                    \
+                                                  &PL_env_mutex,            \
+                                                  PL_env_mutex_depth,       \
+                                                  1);
```
Comment (jkeenan) [inline perl.h]: Why does `define ENV_LOCK` end with a semicolon when the other defines do not?

Comment (khwilliamson) [reply perl.h]: because it is a mistake

File: intrpvar.h [hunk 2]
```diff
@@ -902,6 +902,8 @@ PERLVAR(I, regex_padav,   AV *)		/* All regex objects, indexed via the
 PERLVAR(I, stashpad,    HV **)		/* for CopSTASH */
 PERLVARI(I, stashpadmax, PADOFFSET, 64)
 PERLVARI(I, stashpadix, PADOFFSET, 0)
+PERLVARI(I, env_mutex_depth, int, 0)     /* Emulate general semaphore */
+PERLVARI(I, env_mutex_readers, int, 0)
```
Comment (tonycoz) [inline intrpvar.h]: Where is `PL_env_mutex_readers used`?

I see the copy in perl_clone_using(), but don't see it anywhere else, nor do I see any token pasting in the definitions of PERL_REENTRANT_(LOCK|UNLOCK) that I think would produce it.

Comment (khwilliamson) [reply intrpvar.h]: Those were from a later commit sneaking into this one

**Review by tonycoz — APPROVED** [review]

**Pull request merged**